servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLRequestHandler.kt

```diff
@@ -16,37 +16,134 @@
 import com.expediagroup.graphql.generator.execution.GraphQLContext
+import com.expediagroup.graphql.server.extensions.isDataLoaderInstrumentation
+import com.expediagroup.graphql.server.extensions.isMutation
 import com.expediagroup.graphql.server.extensions.toExecutionInput
 import com.expediagroup.graphql.server.extensions.toGraphQLError
 import com.expediagroup.graphql.server.extensions.toGraphQLKotlinType
 import com.expediagroup.graphql.server.extensions.toGraphQLResponse
+import com.expediagroup.graphql.server.types.GraphQLBatchRequest
+import com.expediagroup.graphql.server.types.GraphQLBatchResponse
 import com.expediagroup.graphql.server.types.GraphQLRequest
 import com.expediagroup.graphql.server.types.GraphQLResponse
+import com.expediagroup.graphql.server.types.GraphQLServerRequest
+import com.expediagroup.graphql.server.types.GraphQLServerResponse
+import graphql.ExecutionInput
 import graphql.GraphQL
+import graphql.execution.instrumentation.ChainedInstrumentation
+import graphql.execution.instrumentation.Instrumentation
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
 import kotlinx.coroutines.future.await
+import kotlinx.coroutines.supervisorScope
 
 open class GraphQLRequestHandler(
     private val graphQL: GraphQL,
-    private val dataLoaderRegistryFactory: DataLoaderRegistryFactory? = null
+    private val dataLoaderRegistryFactory: KotlinDataLoaderRegistryFactory? = null
 ) {
 
+    private val dataLoaderInstrumentationType: Class<Instrumentation>? =
+        graphQL.instrumentation?.let { instrumentation ->
+            when {
+                instrumentation is ChainedInstrumentation -> {
+                    instrumentation.instrumentations
+                        .firstOrNull(Instrumentation::isDataLoaderInstrumentation)
+                        ?.javaClass
+                }
+                instrumentation.isDataLoaderInstrumentation() -> instrumentation.javaClass
+                else -> null
+            }
+        }
+
     /**
      * Execute a GraphQL request in a non-blocking fashion.
      * This should only be used for queries and mutations.
      * Subscriptions require more specific server logic and will need to be handled separately.
      */
-    open suspend fun executeRequest(request: GraphQLRequest, context: GraphQLContext? = null, graphQLContext: Map<*, Any> = emptyMap<Any, Any>()): GraphQLResponse<*> {
-        // We should generate a new registry for every request
-        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
-        val executionInput = request.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
+    open suspend fun executeRequest(
+        graphQLRequest: GraphQLServerRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>()
+    ): GraphQLServerResponse =
+        when (graphQLRequest) {
```

Comment:
I think we can simplify this by moving some stuff around


```kotlin
when (graphQLRequest) {
    is GraphQLRequest -> execute(graphQLRequest, context, graphQLContext)
    is GraphQLBatchRequest -> {
        if (graphQLRequest.containsMutation()) {
            executeBatch(graphQLRequest, context, graphQLContext)
        } else {
            executeBatchAsync(graphQLRequest, context, graphQLContext)
        }
    }
}
```

refactored methods

```kotlin
private suspend fun execute(request: GraphQLRequest, context: GraphQLContext?, graphQLContext: Map<*, Any>, dataLoaderRegistry: KotlinDataLoaderRegistry? = dataLoaderRegistryFactory?.generate()): GraphQLResponse<*> {
    val executionInput = request.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
    return try {
        graphQL.executeAsync(executionInput).await().toGraphQLResponse()
    } catch (exception: Exception) {
        val error = exception.toGraphQLError()
        GraphQLResponse<Any?>(errors = listOf(error.toGraphQLKotlinType()))
    }
}

private suspend fun executeBatch(request: GraphQLBatchRequest, context: GraphQLContext?, graphQLContext: Map<*, Any>): GraphQLBatchResponse =
    GraphQLBatchResponse(request.requests.map { execute(it, context, graphQLContext) })

private suspend fun executeBatchAsync(request: GraphQLBatchRequest, context: GraphQLContext?, graphQLContext: Map<*, Any>): GraphQLBatchResponse {
    val (dataLoaderRegistry, batchContext) = batchingContext(request.requests.size)
    val batchGraphQLContext = graphQLContext + batchContext
    return supervisorScope {
        val responses = request.requests.map {
            async {
                execute(it, context, batchGraphQLContext, dataLoaderRegistry)
            }
        }
        GraphQLBatchResponse(responses.awaitAll())
    }
}

private fun batchingContext(batchSize: Int): Pair<KotlinDataLoaderRegistry?, Map<*, Any>> {
    val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
    val context = dataLoaderRegistry?.let {
        when (dataLoaderInstrumentationType) {
            DataLoaderLevelDispatchedInstrumentation::class.java -> mapOf(
                KotlinDataLoaderRegistry::class to dataLoaderRegistry,
                ExecutionLevelDispatchedState::class to ExecutionLevelDispatchedState(batchSize)
            )
            DataLoaderSyncExecutionExhaustedInstrumentation::class.java -> mapOf(
                KotlinDataLoaderRegistry::class to dataLoaderRegistry,
                SyncExecutionExhaustedState::class to SyncExecutionExhaustedState(batchSize, dataLoaderRegistry)
            )
            else -> null
        }
    } ?: emptyMap<Any, Any>()

    return if (context.isEmpty()) {
        dataLoaderRegistry to context
    } else {
        null to context
    }
}
```

*with new extension
```kotlin
fun GraphQLBatchRequest.containsMutation(): Boolean = requests.any { it.query.contains("mutation ") }
```